infer/inference.cc

```diff
@@ -128,17 +128,61 @@ unique_ptr<cfg::CFG> Inference::run(core::Context ctx, unique_ptr<cfg::CFG> cfg)
+                //      want to issue an error.
+                //   2. If the block contains a send of the form <Magic>.<nil-for-safe-navigation>(x),
+                //      we want to issue an UnnecessarySafeNavigationError, extracting
+                //      type-and-origin info from x. (This magic form is inserted by the desugarer
+                //      for a "safe navigation" operation, e.g., `x&.foo`.)
+                //   3. Otherwise, we want to issue a DeadBranchInferencer error, taking the first
+                //      (non-synthetic, non-"T.absurd") instruction in the block as the loc of the
+                //      error.
+                cfg::Instruction *unreachableInstruction = nullptr;
+                core::LocOffsets locForUnreachable;
+                bool dueToSafeNavigation = false;
+
                 for (auto &expr : bb->exprs) {
                     if (expr.value->isSynthetic) {
                         continue;
                     }
                     if (cfg::isa_instruction<cfg::TAbsurd>(expr.value.get())) {
                         continue;
                     }
-                    if (auto e = ctx.beginError(expr.loc, core::errors::Infer::DeadBranchInferencer)) {
+
+                    auto send = cfg::cast_instruction<cfg::Send>(expr.value.get());
+                    if (send != nullptr && send->fun == core::Names::nilForSafeNavigation()) {
+                        unreachableInstruction = expr.value.get();
+                        locForUnreachable = expr.loc;
+                        dueToSafeNavigation = true;
+                        break;
+                    } else if (unreachableInstruction == nullptr) {
+                        unreachableInstruction = expr.value.get();
+                        locForUnreachable = expr.loc;
+                    }
+                }
+
+                if (unreachableInstruction != nullptr) {
+                    auto send = cfg::cast_instruction<cfg::Send>(unreachableInstruction);
+
+                    if (dueToSafeNavigation && send != nullptr) {
+                        if (auto e =
+                                ctx.beginError(locForUnreachable, core::errors::Infer::UnnecessarySafeNavigation)) {
+                            e.setHeader("Used `{}` operator on a receiver which can never be nil", "&.");
+
+                            // Just a failsafe check; args.size() should always be 1.
+                            if (send->args.size() > 0) {
+                                auto ty = current.getAndFillTypeAndOrigin(ctx, send->args[0]);
+                                e.addErrorSection(
+                                    core::ErrorSection(core::ErrorColors::format("Type of receiver is `{}`, from:",
+                                                                                 ty.type->toString(ctx)),
+                                                       ty.origins2Explanations(ctx, current.ownerLoc)));
```

Comment:
I don't like that there's no obvious connection here that `ownerLoc` is being passed along because it happens to be the loc we used to represent uninitialized values.

Instead, can we make a public helper function on Environment that uses `Environment::uninitialized` to project out the uninitialized loc?